### PR TITLE
CI: Add automatic Github Pages deployment

### PR DESCRIPTION
**Pull request by nalinbhardwaj**
Automates Github Pages deployment by making travis commit any changes to master to gh-pages branch, and other branch changes to gh-pages-branch/"original branch name"

Currently, you can see it works on my fork. For example, commits to `master` automatically get pushed to `gh-pages` and deployed on my website http://nibnalin.me/coala-html/#/ . Branch [CI-testing](https://github.com/nalinbhardwaj/coala-html/tree/CI-testing) gets built and pushed to [gh-pages-branch/CI-testing](https://github.com/nalinbhardwaj/coala-html/tree/gh-pages-branch/CI-testing). Note that the original issue mentions using `gh-pages/<branch name>` but that couldn't be done since git would error out due to existence of `gh-pages` branch.

`data` folder used for building the website has to of course now be moved to master(I've done that), and the failing tests are unrelated to my PR, and have already been verified to fail on clean master by @Mixih 

**TODO**: I would require an owner to provide a  deploy key.

Closes https://github.com/coala/coala-html/issues/133

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by yukiisbored** [comment]
deack https://github.com/coala/coala-html/pull/143/commits/454fe033c4fc61d27d1c5b7954c95310a62300a9

**Comment by nalinbhardwaj** [comment]
@yukiisbored Can you elaborate on the 2nd point? Most of the diff comes from putting the ‘data’ folders from gh-pages branch in master. I’ve only modified 2 files, .travis.yml and .ci/deploy.sh .
The reason behind putting data here is to allow easy change of data to use to build gh-pages, otherwise you would earlier have to rebuild it manually on changing data.
Regarding 1, I’ll check that out but currently [coala/community](https://github.com/coala/community) seems to have a similar deployment system. Should that be changed as well? Also, I think I read online that the way I(and community repo) do it is more secure/modular?

**Comment by yukiisbored** [comment]
@nalinbhardwaj About [coala/community](https://github.com/coala/community), no idea, probably didn't know about Travis CI GitHub Pages deployment feature. Anyway about the Data folder, didn't notice those are coala json output since I reviewed from my phone. But, can you confirm deployment go smoothly since it looks like the data folder isn't inside the `public/` folder which won't be part of the deployed site.

**Comment by nalinbhardwaj** [comment]
Yep @yukiisbored deployment is currently smooth and fine with data where it is. If you want however I can move it to some other directory?

Also there’s a problem with the Travis Github pages building, since it’s a yaml file, I don’t think we can do stuff like pushing to custom branch(as I require for this issue) like master pushes to gh-pages but any other branch “xyz” pushes to gh-pages-branch/xyz ?

**Comment by yukiisbored** [comment]
@nalinbhardwaj see https://docs.travis-ci.com/user/deployment/pages/

**Comment by yukiisbored** [comment]
Also, If you want to see an example: https://github.com/yukiisbored/blog-archive/blob/blog/.travis.yml

**Comment by yukiisbored** [comment]
@nalinbhardwaj just asked @jayvdb, it seems SSH deploy keys are preferred so I'm gonna merge this, but I need to setup the deploy keys.

File: .travis.yml [hunk 1]
```diff
@@ -32,6 +39,8 @@ script:
   - python -m pytest --cov
   - codecov
   - docker run --volume=$(pwd):/app --workdir=/app coala/base coala-ci
+  - python3 setup.py install
+  - bash ./.ci/deploy.sh
```
Comment (yukiisbored) [inline .travis.yml]: This should be done on `deploy` phase so failing builds doesn't get deployed.

File: .ci/deploy.sh [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ENCRYPTED_IV_VAR="encrypted_${ENCRYPTION_LABEL}_iv"
+ENCRYPTED_KEY=${!ENCRYPTED_KEY_VAR}
+ENCRYPTED_IV=${!ENCRYPTED_IV_VAR}
+openssl aes-256-cbc -K $ENCRYPTED_KEY -iv $ENCRYPTED_IV -in $TRAVIS_BUILD_DIR/.ci/deploy_key.enc -out ../deploy_key -d
+chmod 600 ../deploy_key
+eval `ssh-agent -s`
+ssh-add ../deploy_key
+
+# Now that we're all set up, we can push.
+git push $SSH_REPO $TARGET_BRANCH
```
Comment (jayvdb) [inline .ci/deploy.sh]: No EOL at EOF (copying ... :P)

**Comment by yukiisbored** [comment]
Marking this as blocked since the builds are failing.

File: data/Constants.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+	"data":"data",
```
Comment (jayvdb) [inline data/Constants.json]: please add JSONBear to `.coafile`

File: data/roothome [hunk 4]
```diff
@@ -0,0 +1 @@
+/home/tushar/coala
```
Comment (jayvdb) [inline data/roothome]: EOL at EOF

File: .travis.yml [hunk 5]
```diff
@@ -1,12 +1,19 @@
 sudo: required
 dist: trusty
 
+env:
+  global:
+  - ENCRYPTION_LABEL: "79272d66a36b"
```
Comment (jayvdb) [inline .travis.yml]: Yaml lint bear pls

File: .travis.yml (hunk 1 above)
Comment (jayvdb) [inline .travis.yml]: this should be in `deploy:`

Comment (nalinbhardwaj) [reply .travis.yml]: That’s a workaround for now, I’ll do that once tests start passing(as noted in PR description).

**Comment by nalinbhardwaj** [comment]
For future readers, some discussion regarding the decisions can be found in [Zulip](https://coala.zulipchat.com/#narrow/stream/gci/topic/coala-html.20Auto.20deployment).

File: .travis.yml [hunk 6]
```diff
@@ -1,37 +1,46 @@
...
 
+env:
+    global:
+        - ENCRYPTION_LABEL: "79272d66a36b"
+        - COMMIT_AUTHOR_EMAIL: "travis@coala.io"
+
 language: python
 node_js:
-  - "4.1"
+    - "4.1"
```
Comment (yukiisbored) [inline .travis.yml]: Hey this is not needed, it just makes some noise on the diff.

File: .travis.yml [hunk 7]
```diff
@@ -1,37 +1,46 @@
...
 python:
-  - "3.4"
+    - "3.4"
```
Comment (yukiisbored) [inline .travis.yml]: Hey this is not needed, it just makes some noise on the diff.

File: .travis.yml [hunk 8]
```diff
@@ -1,37 +1,46 @@
...
 
 services: docker
 
 cache:
-  pip: true
-  directories:
-    - node_modules
-    - coalahtml/_coalahtml/bower_components
+    pip: true
```
Comment (yukiisbored) [inline .travis.yml]: Hey this is not needed, it just makes some noise on the diff.

File: .travis.yml [hunk 9]
```diff
@@ -1,37 +1,46 @@
...
-  - npm install
-  - cd coalahtml/_coalahtml
-  - bower install
-  - cd -
-  # colorama needs to be upgraded to fix error when installed colorama
-  # version and required are different.
-  - pip install colorama==0.3.7
-  - pip install -r requirements.txt
-  - pip install -r test-requirements.txt
+    - npm install - g bower
```
Comment (yukiisbored) [inline .travis.yml]: Hey this is not needed, it just makes some noise on the diff. 
This breaks CI as well due to that space between - and g

File: .travis.yml [hunk 10]
```diff
@@ -1,37 +1,46 @@
...
+    - npm install
+    - cd coalahtml/_coalahtml
+    - bower install
+    - cd -
+    # colorama needs to be upgraded to fix error when installed colorama
+    # version and required are different.
+    - pip install colorama == 0.3.7
```
Comment (yukiisbored) [inline .travis.yml]: Breaks CI, due to that space between packaage and version

File: .travis.yml [hunk 11]
```diff
@@ -1,37 +1,46 @@
...
+    - pip install - r requirements.txt
```
Comment (yukiisbored) [inline .travis.yml]: Break CI due to that space between the flag and the indicator

File: .travis.yml [hunk 12]
```diff
@@ -1,37 +1,46 @@
...
+    - pip install - r test-requirements.txt
```
Comment (yukiisbored) [inline .travis.yml]: Break CI due to that space between the flag and the indicator

File: .travis.yml [hunk 13]
```diff
@@ -1,37 +1,46 @@
...
+    - codecov
+    - docker run - -volume =$(pwd): / app - -workdir = /app coala/base coala-ci
+
+deploy:
+    - python3 setup.py install
+    - bash ./.ci/deploy.sh
 
 notifications:
-  email: false
+    email: false
```
Comment (yukiisbored) [inline .travis.yml]: Noise, not needed

File: .travis.yml [hunk 14]
```diff
@@ -1,37 +1,46 @@
...
+    - pip install colorama == 0.3.7
+    - pip install - r requirements.txt
+    - pip install - r test-requirements.txt
 
 script:
-  - npm test
-  - python -m pytest --cov
-  - codecov
-  - docker run --volume=$(pwd):/app --workdir=/app coala/base coala-ci
+    - npm test
```
Comment (yukiisbored) [inline .travis.yml]: Not needed.

File: .travis.yml (hunk 6 above)
Comment (nalinbhardwaj) [reply .travis.yml]: Ooh, I just ran YAMLLintBear and applied patch to it without checking anything. Will fix everything.

**Comment by yukiisbored** [comment]
Sadly we can't merge it, master is broken :(

File: .ci/deploy.sh [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+SOURCE_BRANCH="master"
+TARGET_BRANCH="gh-pages"
+
+# Pull requests shouldn't try to deploy
+if [ "$TRAVIS_PULL_REQUEST" != "false" ]; then
+    echo "Skipping deploy."
+    exit 0
+fi
+
+# Commits to other branches should deploy to their respective gh-pages/<name>
```
Comment (blazeu) [inline .ci/deploy.sh]: Should be `gh-pages-branch/<name>` ?

File: .travis.yml [hunk 16]
```diff
@@ -31,7 +36,11 @@ script:
   - npm test
   - python -m pytest --cov
   - codecov
-  - docker run --volume=$(pwd):/app --workdir=/app coala/base coala-ci
+  - docker run --volume =$(pwd): / app --workdir = /app coala/base coala-ci
```
Comment (blazeu) [inline .travis.yml]: I think you shouldn't change this?
See https://travis-ci.org/coala/coala-html/builds/311194465#L686

Comment (yukiisbored) [reply .travis.yml]: @nalinbhardwaj

Comment (yukiisbored) [reply .travis.yml]: Just noticed the command is parsed like YAML

File: .rultor.yml [hunk 17]
```diff
@@ -1,7 +1,7 @@
 merge:
-  fast-forward: only
-  script:
-    - bash .ci/check_maintainership.sh
+    fast-forward: only
```
Comment (yukiisbored) [inline .rultor.yml]: Please undo this.

File: .rultor.yml [hunk 18]
```diff
@@ -1,7 +1,7 @@
...
+    script:
+        - bash .ci/check_maintainership.sh
 
 decrypt:
-  rultor_github_secrets.sh: "repo/.ci/rultor_github_secrets.sh.asc"
+    rultor_github_secrets.sh: "repo/.ci/rultor_github_secrets.sh.asc"
```
Comment (yukiisbored) [inline .rultor.yml]: This as well.

File: .ci/deploy.sh (hunk 15 above)
Comment (nalinbhardwaj) [reply .ci/deploy.sh]: yep, typo.

**Comment by yukiisbored** [comment]
ack https://github.com/coala/coala-html/pull/143/commits/70dfa009d4c8637d44639b36698794a06f8bcb2c

**Comment by nalinbhardwaj** [comment]
Umm, @yukiisbored I think you forgot about the TODO from my PR description? That encrypted deploy key won’t work for this repo?

**Pull request merged**